### ISSUE TITLE
Switch to GitHub Actions

### ISSUE THREAD

**Issue by morganwillcock** [adventuregamestudio/ags-manual#126]
This is already tested separately here on the master branch here:
https://github.com/morganwillcock/ags-manual

As best I can tell it should match what exists at the moment, the main benefits being:
* this removes the need for additional bot accounts to interact with GitHub services
* this moves all builds into a single service
* ...which enables checks that builds come out the same on each platform
* builds are pretty fast

I think that just the build checks (they already revealed some issues) and removal of bot accounts  is reason enough to switch, but the downsides are:
* GitHub Actions itself seem to be fundamentally limited in its design, to the point that the matrix system is not usable if you want to have job outputs, and workflows need to be duplicated to react to different events (e.g. a PR vs a release)
* interactions with other GitHub services actually seems more problematic than when using other CI systems, because the Actions for basic operations like release upload and deployment seem to have a lack of development or do not exist (creating a lot of confusion about which are 'official', which are abandoned, and which actually work)

To try and avoid continual copy and paste errors I've added some templates to the workflows directory so that changes are made in a single place (`make clean && make && git diff` will show you what you've changed). The use of the templating allowed an easy way to create a second workflow that just handles the release process, so there is now a split in the builds between testing that the build system actually works and building the public copies. It also works around the limitations of the job matrix to allow data to be passed directly between jobs, which is what enables the build output to be compared without being forced to re-download all of the build artifacts.

The only functional change would be to the release process, because the relevant Action wants to create the release to pass its reference for the asset upload. So creating and pushing a tag that begins with a 'v' is now how you would create a release.

I don't think I can submit as a PR since the current repository names would need changing for the result to be valid, so I thought to open this issue instead and if no-one objects then I will push the changes directly and update the repository names at the same time.

**Comment by ericoporto** [comment]
I think it's best to move to GitHub Actions too for the manual - much better integration with GitHub! I also found your usage of m4 super interesting.

**Comment by morganwillcock** [comment]
> I think it's best to move to GitHub Actions too for the manual - much better integration with GitHub! I also found your usage of m4 super interesting.

The templating isn't ideal but it actually made the individual builds have fewer steps than if the all jobs had been defined through the matrix. I think it is likely that at some point the matrix issues will be addressed, and perhaps the build process itself will also become simpler.

> I have some questions but they are more me trying to understand how the actions work... What is 2-9-1?

To be able to have jobs using different OS configurations and different Pandoc versions the names are generated as `build-$OS-$SHELL-$PANDOCVERSION`, but GitHub Actions doesn't consider it valid to have '.' characters in the job name (I think because the data selection expressions use '.' as part of the syntax), so I converted them to '-' to end up with job names which are valid.

So '2-9-1' means that it came from Pandoc version '2.9.1'.

> And where/when is GitHub pages built and uploaded?

It is through a separate workflow which only runs when pushing a tag that starts with a 'v'.
https://github.com/morganwillcock/ags-manual/blob/master/.github/workflows/release.yml

This creates a release for the same tag, uploads the CHM file and wiki source as release assets, and deploys the HTML version to GitHub Pages. This sequence was very limited by the working Actions that I could find. For example, the asset upload Action originates from GitHub, but is not officially maintained, and requires creation of the release as a pre-requisite.

**Comment by ericoporto** [comment]
I am sure GitHub will eventually improve actions, it feels they are moving carefully to filter the good things from Pipelines/VSTS but making it nicer, but I like the solutions you came up with. I also liked the artifacts on the pipeline, very nice to quickly inspect what happened.

**Comment by ericoporto** [comment]
Lol, of course they did. If you want, I do have it backed up - I need to look for, but I do have it somewhere.

**Comment by morganwillcock** [comment]
There are pre-built chmcmd binaries in the Free Pascal releases but they are buried within large archives with no way to efficiently download them. The Linux one is in a large uncompressed tar file, then there is another tar file inside that. The Windows one is inside their InnoSetup installer archive. I've not checked macOS yet, but likely there will be a build somewhere within the dmg image.

I've been trying to think of the best way to get them during the build, legally I don't think there would be a problem (effectively it would be a partial copy of what is officially distributed and the licence and source could be bundled with it) but I'm still not sure on the best place to put it. Possibly a repository just to host it (and possibly upload to GitHub Pages) would be OK, since it would effectively be a disposable copy that wouldn't pollute anything else.

I've only had a very quick look, but it seems the CHM file produced by chmcmd is working correctly. The latest version doesn't have the problem with internal page links being flagged as broken, in-fact it flagged up that one of the image links was broken in the Markdown source.

**Comment by morganwillcock** [comment]
The generated search index (what you use from the 'Search' tab, not the 'Index' tab) seems to be much smaller when using the chmcmd build, so instead I've switched the download links for the Microsoft CHM compiler for archive.org versions.

I happened to already distrust the download and store the checksum, so the installation step will fail if someone tampers with the archive.org copy, and right now it passes the verfication check. I don't think this would count as distribution, so legally it should be fine.

So unless I spot any further issues I'll commit all changes in a few days.

**Comment by ericoporto** [comment]
I think HTML Help Workshop from archive.org is fine but if you want to download from GitHub, you can create a branch for storing it on the repository itself and once it's not used anymore - say we switched for offline pages read using a webview in the Editor in the future - the branch can be removed, deleting the file from git history. Alternatively upload it in an issue or Release and pull from there. 

I don't remember, but I think HTML Help Workshop size is relatively small.

**Comment by morganwillcock** [comment]
I'm not sure if it would be legal to host a copy of the compiler. The Chocolatey package maintainer says he is speaking to them about what to do, I've suggested that they also pull it from archive.org but it is questionable whether caching that would be classed as redistribution. The Free Pascal compiler probably needs minimal changes to completely replace or surpass the Microsoft one, so if any further issues come up I think that is probably the path to follow. In the back of my mind I am also considering that documentation is likely going to have to be branched along with large design changes, and I've always thought it best that the scripting implementation supplies doc strings to stay in-sync with the implementation.

I've switched the CI to use GitHub Actions now, although the commit is marked as a failure because it still notified AppVeyor. I've removed the webhooks for AppVeyor and Travis CI now too.

I have a second change that I will open as a Pull Request to check what happens when one is opened.

Closed via https://github.com/adventuregamestudio/ags-manual/commit/c7e8dcdf67b55fffaf913adfb193bf780f9cc1db
